File: src/arithmetic/grader.py
from __future__ import annotations

import re
from dataclasses import dataclass
from pathlib import Path

from .expression import parse_fraction
from .parser import ExpressionParser

EXERCISE_LINE = re.compile(r"^\s*(\d+)\.\s*(.*?)\s*=\s*$")
ANSWER_LINE = re.compile(r"^\s*(\d+)\)\s*(.*?)\s*$")


@dataclass(frozen=True, slots=True)
class GradeResult:
    correct: tuple[int, ...]
    wrong: tuple[int, ...]

    def format(self) -> str:
        return self._line("Correct", self.correct) + "\n" + self._line("Wrong", self.wrong) + "\n"

    @staticmethod
    def _line(label: str, numbers: tuple[int, ...]) -> str:
        return f"{label}: {len(numbers)} ({', '.join(map(str, numbers))})"


class Grader:
    def __init__(self) -> None:
        self._parser = ExpressionParser()

    def grade(self, exercise_file: Path, answer_file: Path) -> GradeResult:
        answers: dict[int, str] = {}
        for line in answer_file.read_text(encoding="utf-8-sig").splitlines():
            if not line.strip():
                continue
            match = ANSWER_LINE.fullmatch(line)
            if match:
                answers[int(match.group(1))] = match.group(2).strip()

        correct: list[int] = []
        wrong: list[int] = []
        for line in exercise_file.read_text(encoding="utf-8-sig").splitlines():
            if not line.strip():
                continue
            match = EXERCISE_LINE.fullmatch(line)
            if not match:
                raise ValueError(f"invalid exercise line: {line}")
            number = int(match.group(1))
            try:
                expected = self._parser.parse(match.group(2)).value
                supplied = answers.get(number)
                if supplied is not None and parse_fraction(supplied) == expected:
                    correct.append(number)
                else:
                    wrong.append(number)
            except (ArithmeticError, ValueError):
                wrong.append(number)
        return GradeResult(tuple(correct), tuple(wrong))

File: src/main/java/com/huahetai/arithmetic/Grader.java
package com.huahetai.arithmetic;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class Grader {
    private static final Pattern EXERCISE = Pattern.compile("^\\s*(\\d+)\\.\\s*(.*?)\\s*=\\s*$");
    private static final Pattern ANSWER = Pattern.compile("^\\s*(\\d+)\\)\\s*(.*?)\\s*$");
    private final ExpressionParser parser = new ExpressionParser();

    public GradeResult grade(Path exerciseFile, Path answerFile) throws IOException {
        Map<Integer, String> answers = parseNumberedLines(answerFile, ANSWER);
        List<Integer> correct = new ArrayList<>();
        List<Integer> wrong = new ArrayList<>();
        for (String line : Files.readAllLines(exerciseFile, StandardCharsets.UTF_8)) {
            if (line.isBlank()) continue;
            Matcher matcher = EXERCISE.matcher(line);
            if (!matcher.matches()) throw new IllegalArgumentException("invalid exercise line: " + line);
            int number = Integer.parseInt(matcher.group(1));
            String supplied = answers.get(number);
            try {
                Rational expected = parser.parse(matcher.group(2)).value();
                if (supplied != null && expected.equals(Rational.parse(supplied))) correct.add(number);
                else wrong.add(number);
            } catch (RuntimeException exception) {
                wrong.add(number);
            }
        }
        return new GradeResult(correct, wrong);
    }

    private Map<Integer, String> parseNumberedLines(Path file, Pattern pattern) throws IOException {
        Map<Integer, String> result = new HashMap<>();
        for (String line : Files.readAllLines(file, StandardCharsets.UTF_8)) {
            if (line.isBlank()) continue;
            Matcher matcher = pattern.matcher(line);
            if (matcher.matches()) result.put(Integer.parseInt(matcher.group(1)), matcher.group(2).trim());
        }
        return result;
    }
}
